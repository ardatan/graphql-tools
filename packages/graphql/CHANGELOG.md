# @graphql-tools/graphql

## 0.0.1

### Patch Changes

- [#4688](https://github.com/ardatan/graphql-tools/pull/4688) [`72dadfe5`](https://github.com/ardatan/graphql-tools/commit/72dadfe558a18e6ac41da28f5965eb740f4f368e) Thanks [@saihaj](https://github.com/saihaj)! - initial release
