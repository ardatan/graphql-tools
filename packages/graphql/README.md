## `@envelop/graphql`

Fork of GraphQL.js
