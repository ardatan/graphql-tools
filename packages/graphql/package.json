{
  "name": "@graphql-tools/graphql",
  "version": "0.0.1",
  "author": "Saihajpreet Singh <saihajpreet.singh@gmail.com>",
  "license": "MIT",
  "sideEffects": false,
  "repository": {
    "type": "git",
    "url": "https://github.com/ardatan/graphql-tools.git",
    "directory": "packages/graphql"
  },
  "keywords": [
    "gql",
    "graphql",
    "typescript"
  ],
  "main": "dist/cjs/index.js",
  "module": "dist/esm/index.js",
  "exports": {
    ".": {
      "require": {
        "types": "./dist/typings/index.d.cts",
        "default": "./dist/cjs/index.js"
      },
      "import": {
        "types": "./dist/typings/index.d.ts",
        "default": "./dist/esm/index.js"
      },
      "default": {
        "types": "./dist/typings/index.d.ts",
        "default": "./dist/esm/index.js"
      }
    },
    "./*": {
      "require": {
        "types": "./dist/typings/*.d.cts",
        "default": "./dist/cjs/*.js"
      },
      "import": {
        "types": "./dist/typings/*.d.ts",
        "default": "./dist/esm/*.js"
      },
      "default": {
        "types": "./dist/typings/*.d.ts",
        "default": "./dist/esm/*.js"
      }
    },
    "./package.json": "./package.json"
  },
  "typings": "dist/typings/index.d.ts",
  "typescript": {
    "definition": "dist/typings/index.d.ts"
  },
  "devDependencies": {
    "typescript": "4.8.4"
  },
  "buildOptions": {
    "input": "./src/index.ts"
  },
  "publishConfig": {
    "directory": "dist",
    "access": "public"
  },
  "type": "module"
}
