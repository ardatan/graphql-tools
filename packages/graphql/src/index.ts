export * from './error/index.js';
export * from './execution/index.js';
export * from './graphql.js';
export * from './language/index.js';
export * from './type/index.js';
export * from './utilities/index.js';
export * from './validation/index.js';
export * from './version.js';
export * from './subscription/index.js';
