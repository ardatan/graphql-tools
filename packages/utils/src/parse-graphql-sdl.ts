import {
  DocumentNode,
  Kind,
  ASTNode,
  parse,
  Source as GraphQLSource,
  visit,
  isTypeSystemDefinitionNode,
  StringValueNode,
  print,
} from 'graphql';
import { dedentBlockStringValue, getLeadingCommentBlock } from './comments.js';
import { GraphQLParseOptions } from './Interfaces.js';

export function parseGraphQLSDL(location: string | undefined, rawSDL: string, options: GraphQLParseOptions = {}) {
  let document: DocumentNode;

  try {
    if (options.commentDescriptions && rawSDL.includes('#')) {
      document = transformCommentsToDescriptions(rawSDL, options);

      // If noLocation=true, we need to make sure to print and parse it again, to remove locations,
      // since `transformCommentsToDescriptions` must have locations set in order to transform the comments
      // into descriptions.
      if (options.noLocation) {
        document = parse(print(document), options);
      }
    } else {
      document = parse(new GraphQLSource(rawSDL, location), options);
    }
  } catch (e: any) {
    if (e.message.includes('EOF') && rawSDL.replace(/(\#[^*]*)/g, '').trim() === '') {
      document = {
        kind: Kind.DOCUMENT,
        definitions: [],
      };
    } else {
      throw e;
    }
  }

  return {
    location,
    document,
  };
}

export function transformCommentsToDescriptions(sourceSdl: string, options: GraphQLParseOptions = {}): DocumentNode {
  const parsedDoc = parse(sourceSdl, {
    ...options,
    noLocation: false,
  });
  const modifiedDoc = visit(parsedDoc, {
    leave: (node: ASTNode) => {
      if (isDescribable(node)) {
        const rawValue = getLeadingCommentBlock(node);

        if (rawValue !== undefined) {
          const commentsBlock = dedentBlockStringValue('\n' + rawValue);
          const isBlock = commentsBlock.includes('\n');

          if (!node.description) {
            return {
              ...node,
              description: {
                kind: Kind.STRING,
                value: commentsBlock,
                block: isBlock,
              },
            };
          } else {
            return {
              ...node,
              description: {
                ...node.description,
                value: node.description.value + '\n' + commentsBlock,
                block: true,
              },
            };
          }
        }
      }
    },
  });

  return modifiedDoc;
}

type DiscriminateUnion<T, U> = T extends U ? T : never;
type DescribableASTNodes = DiscriminateUnion<
  ASTNode,
  {
    description?: StringValueNode;
  }
>;

export function isDescribable(node: ASTNode): node is DescribableASTNodes {
  return (
    isTypeSystemDefinitionNode(node) ||
    node.kind === Kind.FIELD_DEFINITION ||
    node.kind === Kind.INPUT_VALUE_DEFINITION ||
    node.kind === Kind.ENUM_VALUE_DEFINITION
  );
}
