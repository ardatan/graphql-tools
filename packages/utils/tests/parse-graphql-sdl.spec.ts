import { Kind, ObjectTypeDefinitionNode, print } from 'graphql';
import { parseGraphQLSDL, transformCommentsToDescriptions } from '../src/parse-graphql-sdl.js';

describe('parse sdl', () => {
  describe('parseGraphQLSDL', () => {
    it('Should work correctly with empty document (everything is commented out)', () => {
      const doc = /* GraphQL */ `
        #query test {
        #  id
        #}
      `;

      expect(parseGraphQLSDL('a.graphql', doc).document).toEqual({
        kind: Kind.DOCUMENT,
        definitions: [],
      });
    });
  });

  describe('comment parsing', () => {
    const ast = /* GraphQL */ `
      # test type comment
      type Type {
        # test field comment
        f1: String!
        # Line 1
        #Line 2
        f2: String!
        # Line 1
        "line 2"
        f3: String!
      }

      # test extension
      extend type Type {
        # test extension field comment
        f4: String!
      }

      type OtherType implements Node {
        id: ID!
        f: String!
      }

      # input test
      input Input {
        # Input field test
        f: String
      }

      # Enum test
      enum Enum {
        # Enum value test
        V1
        V2
      }

      # Union test
      union Union = Type | OtherType

      # Interface test
      interface Node {
        id: ID!
      }

      # Custom scalar test
      scalar Date
    `;

    it('should transform comments to descriptions correctly on all available nodes', () => {
      const transformed = transformCommentsToDescriptions(ast);
      const printed = print(transformed).trim();

      expect(printed).toMatchInlineSnapshot(`
        ""test type comment"
        type Type {
          "test field comment"
          f1: String!
          """
           Line 1
          Line 2
          """
          f2: String!
          """
          line 2
          Line 1
          """
          f3: String!
        }

        extend type Type {
          "test extension field comment"
          f4: String!
        }

        type OtherType implements Node {
          id: ID!
          f: String!
        }

        "input test"
        input Input {
          "Input field test"
          f: String
        }

        "Enum test"
        enum Enum {
          "Enum value test"
          V1
          V2
        }

        "Union test"
        union Union = Type | OtherType

        "Interface test"
        interface Node {
          id: ID!
        }

        "Custom scalar test"
        scalar Date"
      `);
    });

    it('should transform comments to descriptions correctly on all available nodes with noLocation=true', () => {
      const transformed = parseGraphQLSDL('test.graphql', ast, {
        noLocation: true,
        commentDescriptions: true,
      });
      const type = transformed.document.definitions.find(
        (d): d is ObjectTypeDefinitionNode => 'name' in d && d.name?.value === 'Type',
      );
      expect(type?.description?.value).toBe('test type comment');
      expect(type?.loc).not.toBeDefined();
      const printed = print(transformed.document).trim();
      expect(printed).toMatchInlineSnapshot(`
        ""test type comment"
        type Type {
          "test field comment"
          f1: String!
          """
           Line 1
          Line 2
          """
          f2: String!
          """
          line 2
          Line 1
          """
          f3: String!
        }

        extend type Type {
          "test extension field comment"
          f4: String!
        }

        type OtherType implements Node {
          id: ID!
          f: String!
        }

        "input test"
        input Input {
          "Input field test"
          f: String
        }

        "Enum test"
        enum Enum {
          "Enum value test"
          V1
          V2
        }

        "Union test"
        union Union = Type | OtherType

        "Interface test"
        interface Node {
          id: ID!
        }

        "Custom scalar test"
        scalar Date"
      `);
    });
  });
});
