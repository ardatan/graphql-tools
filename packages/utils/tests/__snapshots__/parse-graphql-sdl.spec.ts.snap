// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse sdl comment parsing should transform comments to descriptions correctly on all available nodes 1`] = `
"\\"test type comment\\"
type Type {
  \\"test field comment\\"
  f1: String!
  \\"\\"\\"
   Line 1
  Line 2
  \\"\\"\\"
  f2: String!
  \\"\\"\\"
  line 2
  Line 1
  \\"\\"\\"
  f3: String!
}

extend type Type {
  \\"test extension field comment\\"
  f4: String!
}

type OtherType implements Node {
  id: ID!
  f: String!
}

\\"input test\\"
input Input {
  \\"Input field test\\"
  f: String
}

\\"Enum test\\"
enum Enum {
  \\"Enum value test\\"
  V1
  V2
}

\\"Union test\\"
union Union = Type | OtherType

\\"Interface test\\"
interface Node {
  id: ID!
}

\\"Custom scalar test\\"
scalar Date"
`;

exports[`parse sdl comment parsing should transform comments to descriptions correctly on all available nodes with noLocation=true 1`] = `
"\\"test type comment\\"
type Type {
  \\"test field comment\\"
  f1: String!
  \\"\\"\\"
   Line 1
  Line 2
  \\"\\"\\"
  f2: String!
  \\"\\"\\"
  line 2
  Line 1
  \\"\\"\\"
  f3: String!
}

extend type Type {
  \\"test extension field comment\\"
  f4: String!
}

type OtherType implements Node {
  id: ID!
  f: String!
}

\\"input test\\"
input Input {
  \\"Input field test\\"
  f: String
}

\\"Enum test\\"
enum Enum {
  \\"Enum value test\\"
  V1
  V2
}

\\"Union test\\"
union Union = Type | OtherType

\\"Interface test\\"
interface Node {
  id: ID!
}

\\"Custom scalar test\\"
scalar Date"
`;
